Refresh existing Exception Handling guidelines (#15)
input-output-hk/adrestia: process/eng-practices/Exceptions-Haskell.md

---
tags:
  - needs/review
  - needs/writing
---
# Exceptions in Haskell

In `cardano-wallet`, we currently use the following conventions for dealing with exceptions:

1. Avoid partial functions.\

>     Example: Do not use `head`, but use a pattern match, or consider using functions from `Safe` module, such as `headMay`. Considering all corner cases does catch bugs.

Reviewer: Here's another example that is a little unexpected: `Data.Text.Encoding.decodeUtf8`. It crashes with invalid input.

2. Pure code vs `IO` actions
    * In pure code, use `error` instead of `undefined` if really necessary. In this case, please also add a `HasCallStack` constraint so that a backtrace can be shown when the impossible happens.

>     * In `IO`, use `throwIO` to throw a regular exception.

Reviewer: I do like to recommend `UnliftIO.Exception.throwString` to lazy developers as a more user-friendly alternative to `error`.

3. Import exception utilities from `UnliftIO.Exception` module rather than `Control.Exception`; these utilities have better default behavior for asynchronous exceptions.

4. If you desire to log an exception, use the tracing framework.

5. Be specific when catching exceptions.\
    For example, if you only want to catch errors when reading a file which doesn't exist use `tryJust`, `ioeGetErrorType`, `isDoesNotExistErrorType`, etc.

6. Be aware that `IOException`s can have different forms on different platforms, especially with Windows and networking code. You need to test your exception handling on Windows.

7. Checked vs. unchecked exceptions
    * Please do not mix `IO (Either e a)` or `ExceptT` with unchecked exceptions.

>     * If using checked exceptions, do not add an `Exception` instance to your error type, because this will ensure that it cannot be thrown as an unchecked exception.

Reviewer: ```suggestion
    * If using checked exceptions, do not add an `Exception` instance to your error type. Otherwise it can be thrown as an unchecked exception, which is really confusing.
```
